File: src-tauri/src/notifications/trigger_tester.rs
//! `trigger_tester` — tests a trigger against all historical sessions with
//! safety limits. Ported from `internal/notifications/trigger_tester.go` (W14).

use std::path::{Path, PathBuf};
use std::time::{Duration, Instant, SystemTime};

use crate::config::state::types::NotificationTrigger;
use crate::discovery::path_decoder::get_projects_base_path;
use crate::discovery::project_scanner::scan_projects;
use crate::discovery::subproject_registry::SubprojectRegistry;
use crate::parsing::session_parser::parse_jsonl_file;
use crate::watcher::resolve_claude_dir;

use super::error_detector::detect_errors_with_trigger;
use super::types::{DetectedError, TriggerTestResult};

const MAX_ERRORS: usize = 50;
const MAX_TOTAL_COUNT: usize = 10_000;
const TIMEOUT: Duration = Duration::from_secs(30);

struct TestState {
    errors: Vec<DetectedError>,
    total_count: usize,
    truncated: bool,
    start_time: Instant,
    effective_limit: usize,
}

impl TestState {
    fn should_stop(&self) -> bool {
        self.errors.len() >= self.effective_limit
            || self.start_time.elapsed() > TIMEOUT
            || self.total_count >= MAX_TOTAL_COUNT
    }

    fn is_safety_limit(&self) -> bool {
        self.start_time.elapsed() > TIMEOUT || self.total_count >= MAX_TOTAL_COUNT
    }
}

fn empty_result() -> TriggerTestResult {
    TriggerTestResult {
        total_count: 0,
        errors: Vec::new(),
        truncated: None,
    }
}

/// Tests a trigger against all historical session data with hard safety limits
/// (50 errors or caller `limit`, 10k total scanned, 30s wall clock).
pub fn test_trigger(trigger: &NotificationTrigger, limit: Option<usize>) -> TriggerTestResult {
    let effective_limit = match limit {
        Some(l) if l < MAX_ERRORS => l,
        _ => MAX_ERRORS,
    };

    let mut state = TestState {
        errors: Vec::new(),
        total_count: 0,
        truncated: false,
        start_time: Instant::now(),
        effective_limit,
    };

    let claude_dir = match resolve_claude_dir() {
        Some(dir) => dir,
        None => return empty_result(),
    };

    let projects_dir = get_projects_base_path(&claude_dir);
    let mut registry = SubprojectRegistry::new();
    let projects = match scan_projects(&projects_dir, &mut registry) {
        Ok(p) => p,
        Err(_) => return empty_result(),
    };

    'outer: for project in &projects {
        if state.should_stop() {
            if state.is_safety_limit() {
                state.truncated = true;
            }
            break;
        }

        let project_dir = projects_dir.join(&project.id);
        let session_files = match list_jsonl_files(&project_dir) {
            Ok(f) => f,
            Err(_) => continue,
        };

        for file_path in &session_files {
            if state.should_stop() {
                if state.is_safety_limit() {
                    state.truncated = true;
                }
                break 'outer;
            }

            let messages = match parse_jsonl_file(file_path) {
                Ok((msgs, _)) => msgs,
                Err(_) => continue,
            };

            let session_id = file_path
                .file_stem()
                .map(|s| s.to_string_lossy().into_owned())
                .unwrap_or_default();

            let session_errors = detect_errors_with_trigger(
                &messages,
                trigger,
                &session_id,
                &project.id,
                &file_path.to_string_lossy(),
            );

            let new_total = state.total_count + session_errors.len();
            if new_total >= MAX_TOTAL_COUNT {
                state.total_count = MAX_TOTAL_COUNT;
                state.truncated = true;
            } else {
                state.total_count = new_total;
            }

            for e in session_errors {
                if state.errors.len() >= state.effective_limit {
                    break;
                }
                state.errors.push(e);
            }
        }
    }

    TriggerTestResult {
        total_count: state.total_count as i64,
        errors: state.errors,
        truncated: if state.truncated { Some(true) } else { None },
    }
}

/// Returns `.jsonl` files in `dir`, sorted newest-first by mtime.
fn list_jsonl_files(dir: &Path) -> std::io::Result<Vec<PathBuf>> {
    let mut files: Vec<(PathBuf, SystemTime)> = Vec::new();
    for entry in std::fs::read_dir(dir)? {
        let entry = match entry {
            Ok(e) => e,
            Err(_) => continue,
        };
        let path = entry.path();
        if path.is_dir() || path.extension().and_then(|e| e.to_str()) != Some("jsonl") {
            continue;
        }
        let mtime = match entry.metadata().and_then(|m| m.modified()) {
            Ok(t) => t,
            Err(_) => continue,
        };
        files.push((path, mtime));
    }

    files.sort_by(|a, b| b.1.cmp(&a.1));
    Ok(files.into_iter().map(|(p, _)| p).collect())
}

#[cfg(test)]
#[path = "trigger_tester_tests.rs"]
mod tests;

File: internal/notifications/trigger_tester.go
// trigger_tester.go.
// Tests a trigger against all historical sessions with safety limits.
package notifications

import (
	"os"
	"path/filepath"
	"sort"
	"time"

	"claude-devtools/internal/config"
	"claude-devtools/internal/discovery"
	"claude-devtools/internal/parsing"
)

const (
	maxErrors       = 50
	maxTotalCount   = 10_000
	timeoutDuration = 30 * time.Second
)

type testState struct {
	errors        []DetectedError
	totalCount    int
	truncated     bool
	startTime     time.Time
	effectiveLimit int
}

func (ts *testState) shouldStop() bool {
	return len(ts.errors) >= ts.effectiveLimit ||
		time.Since(ts.startTime) > timeoutDuration ||
		ts.totalCount >= maxTotalCount
}

func (ts *testState) isSafetyLimit() bool {
	return time.Since(ts.startTime) > timeoutDuration || ts.totalCount >= maxTotalCount
}

// TestTrigger tests a trigger against all historical session data.
// Mirrors trigger_tester.rs::test_trigger.
func TestTrigger(trigger *config.NotificationTrigger, limit *int) TriggerTestResult {
	effectiveLimit := maxErrors
	if limit != nil && *limit < maxErrors {
		effectiveLimit = *limit
	}

	state := &testState{
		effectiveLimit: effectiveLimit,
		startTime:      time.Now(),
	}

	claudeDir := resolveClaudeDir()
	if claudeDir == "" {
		return TriggerTestResult{TotalCount: 0, Errors: []DetectedError{}}
	}

	projectsDir := discovery.GetProjectsBasePath(claudeDir)
	registry := discovery.NewSubprojectRegistry()

	projects, err := discovery.ScanProjects(projectsDir, registry)
	if err != nil {
		return TriggerTestResult{TotalCount: 0, Errors: []DetectedError{}}
	}

outer:
	for _, project := range projects {
		if state.shouldStop() {
			if state.isSafetyLimit() {
				state.truncated = true
			}
			break
		}

		projectDir := filepath.Join(projectsDir, project.ID)
		sessionFiles, err := listJSONLFiles(projectDir)
		if err != nil {
			continue
		}

		for _, filePath := range sessionFiles {
			if state.shouldStop() {
				if state.isSafetyLimit() {
					state.truncated = true
				}
				break outer
			}

			messages, _, err := parsing.ParseJSONLFile(filePath)
			if err != nil {
				continue
			}

			sessionID := filepath.Base(filePath)
			if ext := filepath.Ext(sessionID); ext != "" {
				sessionID = sessionID[:len(sessionID)-len(ext)]
			}

			sessionErrors := DetectErrorsWithTrigger(messages, trigger, sessionID, project.ID, filePath)

			newTotal := state.totalCount + len(sessionErrors)
			if newTotal >= maxTotalCount {
				state.totalCount = maxTotalCount
				state.truncated = true
			} else {
				state.totalCount = newTotal
			}

			for _, e := range sessionErrors {
				if len(state.errors) >= state.effectiveLimit {
					break
				}
				state.errors = append(state.errors, e)
			}
		}
	}

	if state.errors == nil {
		state.errors = []DetectedError{}
	}

	var truncated *bool
	if state.truncated {
		t := true
		truncated = &t
	}

	return TriggerTestResult{
		TotalCount: state.totalCount,
		Errors:     state.errors,
		Truncated:  truncated,
	}
}

// resolveClaudeDir mirrors watcher::resolve_claude_dir — returns ~/.claude or "".
func resolveClaudeDir() string {
	home, err := os.UserHomeDir()
	if err != nil {
		return ""
	}
	dir := filepath.Join(home, ".claude")
	if _, err := os.Stat(dir); err != nil {
		return ""
	}
	return dir
}

// listJSONLFiles returns .jsonl files in a directory, sorted newest-first by mtime.
// Mirrors trigger_tester.rs::list_jsonl_files.
func listJSONLFiles(dir string) ([]string, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}

	type fileEntry struct {
		path  string
		mtime time.Time
	}
	var files []fileEntry
	for _, e := range entries {
		if e.IsDir() || filepath.Ext(e.Name()) != ".jsonl" {
			continue
		}
		info, err := e.Info()
		if err != nil {
			continue
		}
		files = append(files, fileEntry{
			path:  filepath.Join(dir, e.Name()),
			mtime: info.ModTime(),
		})
	}

	sort.Slice(files, func(i, j int) bool {
		return files[i].mtime.After(files[j].mtime)
	})

	paths := make([]string, len(files))
	for i, f := range files {
		paths[i] = f.path
	}
	return paths, nil
}
